File: backend/src/routes/projects.ts
import { Router } from 'express';
import * as cache from '../util/cache';
import { ProjectsGetResponse, Project } from '../../../shared/api/projects';
import { sequelize } from '../db/index';
import { QueryTypes } from 'sequelize';

const router = Router();

router.get('/', async (req, res) => {
    let projects = cache.get<ProjectsGetResponse>(req.session.user.id, 'projects');
    console.log(projects);
    if(projects) {
        res.send(projects);
        return;
    }
    
    let dbProjects = await sequelize.query<Project>('SELECT "project", SUM("end" - "start") as "totalTime" FROM "TimeEntries" WHERE "userId"=:userId GROUP BY "project"', {
        type: QueryTypes.SELECT,
        replacements: { userId: req.session.user.id },
    });
    // TODO convert postgresql interval

    projects = {
        projects: dbProjects,
    };

    cache.set<ProjectsGetResponse>(req.session.user.id, 'projects', projects);
    res.send(projects);
});

export default router;

File: backend/routes/projects.go
package routes

import (
	"encoding/json"
	"net/http"

	"github.com/peterhellberg/duration"

	"github.com/winded/tyomaa/backend/db"
	"github.com/winded/tyomaa/shared/api"

	"github.com/winded/tyomaa/backend/util"
	"github.com/winded/tyomaa/backend/util/context"

	"github.com/gorilla/mux"
	"github.com/winded/tyomaa/backend/middleware"
)

func ProjectsRoutes(router *mux.Router) {
	const projectsQuery = `SELECT "project" as "name", SUM("end" - "start") as "total_time" FROM time_entries WHERE user_id = ? GROUP BY "project" ORDER BY "project" ASC`

	router.Use(middleware.Authorization)

	router.Path("").Methods(http.MethodGet).HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		defer util.HandleApiError(w)

		ctx := context.Get(r)

		rows, err := db.Instance.Raw(projectsQuery, ctx.User.ID).Rows()
		if err != nil {
			panic(err)
		}
		defer rows.Close()

		var results []api.Project
		for rows.Next() {
			var (
				name string
				t    string
			)
			if err := rows.Scan(&name, &t); err != nil {
				panic(err)
			}

			totalTime, err := duration.Parse(t)
			if err != nil {
				panic(err)
			}

			results = append(results, api.Project{
				Name:      name,
				TotalTime: totalTime,
			})
		}

		json.NewEncoder(w).Encode(api.ProjectsGetResponse{
			Projects: results,
		})
	})
}
